File: using python/367. Valid Perfect Square.py
class Solution:
    def isPerfectSquare(self, num):
        if num == 1:
            return True
        
        low = 1
        high = num
        
        while low <= high:
            mid = (low + high) // 2
            sqr = mid * mid
            
            if sqr == num:
                return True
            elif sqr < num:
                low = mid + 1
            else:
                high = mid - 1
        
        return False

#to run script locally in your system
# myclass=Solution()
# n=5
# print(myclass.isPerfectSquare(n))

File: Using Java/367. Valid Perfect Square.java
class Solution {
    public boolean isPerfectSquare(int num) {
        if (num == 1) {
            return true;
        }
        
        int low = 1;
        int high = num;
        
        while (low <= high) {
            int mid = low + (high - low) / 2;
            long sqr = (long) mid * mid;  // Use long to avoid overflow
            
            if (sqr == num) {
                return true;
            } else if (sqr < num) {
                low = mid + 1;
            } else {
                high = mid - 1;
            }
        }
        
        return false;
    }
}
